### PR TITLE
Cosmos: CFP fixes and port consistency policy bug fix

### PR DESCRIPTION
**Pull request by milismsft**
ChangeFeedProcessor; fixes and extra logging related to the creations of the lease documents.
Port consistency policy bug fix (see https://github.com/Azure/azure-cosmosdb-java/pull/196)
Port test fixes (see https://github.com/Azure/azure-cosmosdb-java/pull/196)

### CODE REVIEW

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java [hunk 1]
```diff
@@ -143,10 +152,11 @@ public DocumentServiceLeaseStore(
                 if (throwable instanceof CosmosClientException) {
                     CosmosClientException e = (CosmosClientException) throwable;
                     if (e.statusCode() == ChangeFeedHelper.HTTP_STATUS_CODE_CONFLICT) {
+                        logger.info("Lease synchronization document was acquired by a different instance");
                         return Mono.just(false);
                     }
                 }
-
+                logger.warn("Unexpected exception thrown", throwable);
```
Comment (kushagraThapar) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java]: Should be debug. Warn is too much of noise.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/LeaseRenewerImpl.java [hunk 2]
```diff
@@ -63,7 +63,11 @@ public LeaseRenewerImpl(Lease lease, LeaseManager leaseManager, Duration leaseRe
             })
             .then()
             .onErrorResume(throwable -> {
-                logger.error("Partition {}: renew lease loop failed.", this.lease.getLeaseToken(), throwable);
+                if (throwable instanceof LeaseLostException) {
+                    logger.warn("Partition {}: renew lease loop failed.", this.lease.getLeaseToken(), throwable);
```
Comment (kushagraThapar) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/LeaseRenewerImpl.java]: let's change these to debug, or error if you want to. But not warn.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java [hunk 3]
```diff
@@ -104,9 +111,11 @@ public DocumentServiceLeaseStore(
                 if (throwable instanceof CosmosClientException) {
                     CosmosClientException e = (CosmosClientException) throwable;
                     if (e.statusCode() == ChangeFeedHelper.HTTP_STATUS_CODE_CONFLICT) {
+                        logger.info("Lease synchronization document was acquired by a different instance");
                         return Mono.just(false);
                     }
                 }
+                logger.warn("Unexpected exception thrown", throwable);
```
Comment (kushagraThapar) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java]: error for this one too.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java [hunk 4]
```diff
@@ -71,15 +76,17 @@ public DocumentServiceLeaseStore(
...
             .map( item -> true)
             .onErrorResume(throwable -> {
                 if (throwable instanceof CosmosClientException) {
                     CosmosClientException e = (CosmosClientException) throwable;
                     if (e.statusCode() == ChangeFeedHelper.HTTP_STATUS_CODE_CONFLICT) {
+                        logger.info("Lease synchronization document was created by a different instance");
                         return Mono.just(true);
                     }
                 }
+                logger.warn("Unexpected exception thrown", throwable);
```
Comment (kushagraThapar) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java]: error for this one and other places.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java (hunk 1 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/LeaseRenewerImpl.java (hunk 2 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/LeaseRenewerImpl.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java (hunk 3 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java (hunk 4 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/DocumentServiceLeaseStore.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java [hunk 5]
```diff
@@ -318,34 +318,25 @@ boolean isGlobalStrongRequest(RxDocumentServiceRequest request, StoreResponse re
...
 
                     return Mono.empty();
                     }).flux();
-        }).repeatWhen(s -> {
-            if (writeBarrierRetryCount.get() == 0) {
-                    return Flux.empty();
+        }).repeatWhen(s -> s.flatMap(x -> {
+            // repeat with a delay
+            if ((ConsistencyWriter.MAX_NUMBER_OF_WRITE_BARRIER_READ_RETRIES - writeBarrierRetryCount.get()) > ConsistencyWriter.MAX_SHORT_BARRIER_RETRIES_FOR_MULTI_REGION) {
+                return Flux.just(0L).delayElements(Duration.ofMillis(ConsistencyWriter.DELAY_BETWEEN_WRITE_BARRIER_CALLS_IN_MS));
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java]: This works but you have some dummy value (`0L`) which makes it a bit hard to read.

please replace
`Flux.just(0L).delayElements(duration)` with `Mono.delay(duration).flux()`.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java [hunk 6]
```diff
@@ -318,34 +318,25 @@ boolean isGlobalStrongRequest(RxDocumentServiceRequest request, StoreResponse re
...
             } else {
-
-                if ((ConsistencyWriter.MAX_NUMBER_OF_WRITE_BARRIER_READ_RETRIES - writeBarrierRetryCount.get()) > ConsistencyWriter.MAX_SHORT_BARRIER_RETRIES_FOR_MULTI_REGION) {
-                    return Flux.just(0L).delayElements(Duration.ofMillis(ConsistencyWriter.DELAY_BETWEEN_WRITE_BARRIER_CALLS_IN_MS));
-                } else {
-                    return Flux.just(0L).delayElements(Duration.ofMillis(ConsistencyWriter.SHORT_BARRIER_RETRY_INTERVAL_IN_MS_FOR_MULTI_REGION));
-                }
+                return Flux.just(0L).delayElements(Duration.ofMillis(ConsistencyWriter.SHORT_BARRIER_RETRY_INTERVAL_IN_MS_FOR_MULTI_REGION));
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java]: here too.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java [hunk 7]
```diff
@@ -248,7 +248,11 @@ public StoreReader(
             for (StoreResult srr : newStoreResults) {
 
                 entity.requestContext.requestChargeTracker.addCharge(srr.requestCharge);
-                BridgeInternal.recordResponse(entity.requestContext.cosmosResponseDiagnostics, entity, srr);
+                try {
+                    BridgeInternal.recordResponse(entity.requestContext.cosmosResponseDiagnostics, entity, srr);
+                } catch (Exception e) {
+                    logger.error("Error occurred while recording response", e);
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java]: Please similar to the original PR, change to "Unexpected failure while recording response"

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java [hunk 8]
```diff
@@ -557,7 +561,11 @@ private ReadReplicaResult createReadReplicaResult(List<StoreResult> responseResu
         });
 
         return storeResultObs.map(storeResult -> {
-            BridgeInternal.recordResponse(entity.requestContext.cosmosResponseDiagnostics, entity, storeResult);
+            try {
+                BridgeInternal.recordResponse(entity.requestContext.cosmosResponseDiagnostics, entity, storeResult);
+            } catch (Exception e) {
+                logger.error("Error occurred while recording response", e);
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java]: Please similar to the original PR, change to "Unexpected failure while recording response"

**Comment by moderakh** [comment]
@milismsft Please run all the tests. We need to run all the tests for checkin. I see only emulator tests above.

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java (hunk 5 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java (hunk 6 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java (hunk 8 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java (hunk 7 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/StoreReader.java]: Fixed

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java [hunk 9]
```diff
@@ -309,7 +309,7 @@ boolean isGlobalStrongRequest(RxDocumentServiceRequest request, StoreResponse re
                     }
 
                     //get max global committed lsn from current batch of responses, then update if greater than max of all batches.
-                    long maxGlobalCommittedLsn = (responses != null || !responses.isEmpty()) ?
+                    long maxGlobalCommittedLsn = (responses != null && !responses.isEmpty()) ?
```
Comment (moderakh) [inline sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java]: please change to this.
```java
                    long maxGlobalCommittedLsn = (responses != null) ?
                        (Long) responses.stream().map(s -> s.globalCommittedLSN).max(ComparatorUtils.NATURAL_COMPARATOR).orElse(0L) :
                        0L;
```

**Comment by moderakh** [comment]
This needs to be ported to v4 @kushagraThapar

File: sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java (hunk 9 above)
Comment (milismsft) [reply sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/directconnectivity/ConsistencyWriter.java]: Fixed

**Pull request merged**